File: apps/api/src/backtest/engine.ts
/**
 * 백테스트 엔진 1단계 — RSI 임계값 전략 (docs/menu/backtest/05)
 */
import * as history from "../data/history.js";

export interface BacktestParams {
  start_date: string;
  end_date: string;
  symbols: string[];
  strategy: { type: string; params?: Record<string, number> };
  initial_capital?: number;
  commission_pct?: number;
}

export interface Trade {
  date: string;
  symbol: string;
  side: "buy" | "sell";
  price: number;
  quantity: number;
  pnl?: number;
  pnl_pct?: number;
}

export interface BacktestResult {
  total_return_pct: number;
  annualized_return_pct: number;
  max_drawdown_pct: number;
  trade_count: number;
  win_rate: number;
  equity_curve: number[];
  trades: Trade[];
  start_date: string;
  end_date: string;
}

const DEFAULT_CAPITAL = 1_000_000;
const DEFAULT_COMMISSION = 0;

export async function runBacktest(params: BacktestParams): Promise<BacktestResult> {
  const { start_date, end_date, symbols, strategy } = params;
  const symbol = symbols?.[0] ?? "AAPL";
  const initialCapital = params.initial_capital ?? DEFAULT_CAPITAL;
  const commissionPct = params.commission_pct ?? DEFAULT_COMMISSION;

  const rows = await history.getHistory(symbol, start_date, end_date);
  if (rows.length === 0) {
    return {
      total_return_pct: 0,
      annualized_return_pct: 0,
      max_drawdown_pct: 0,
      trade_count: 0,
      win_rate: 0,
      equity_curve: [],
      trades: [],
      start_date,
      end_date,
    };
  }

  const type = (strategy?.type as string) || "rsi_threshold";
  const p = strategy?.params ?? {};
  const period = Math.max(2, Number(p.period) || 14);
  const buyBelow = Number(p.buy_below) ?? 30;
  const sellAbove = Number(p.sell_above) ?? 70;

  const closes = rows.map((r) => r.close);
  const rsiValues = history.computeRSI(closes, period);

  const trades: Trade[] = [];
  let position = 0;
  let entryPrice = 0;
  let cash = initialCapital;
  let equity = initialCapital;
  const equityCurve: number[] = [];

  for (let i = 0; i < rows.length; i++) {
    const row = rows[i]!;
    const rsi = rsiValues[i];
    const price = row.close;

    if (rsi == null) {
      equityCurve.push(equity / initialCapital);
      continue;
    }

    if (position === 0 && rsi < buyBelow) {
      const cost = price * (1 + commissionPct / 100);
      const qty = Math.floor(cash / cost);
      if (qty > 0) {
        position = qty;
        entryPrice = price;
        cash -= cost * qty;
        trades.push({ date: row.date, symbol, side: "buy", price, quantity: qty });
      }
    } else if (position > 0 && rsi > sellAbove) {
      const proceeds = price * (1 - commissionPct / 100) * position;
      const pnl = (price - entryPrice) * position;
      const pnlPct = (price / entryPrice - 1) * 100;
      cash += proceeds;
      trades.push({
        date: row.date,
        symbol,
        side: "sell",
        price,
        quantity: position,
        pnl: Math.round(pnl * 100) / 100,
        pnl_pct: Math.round(pnlPct * 100) / 100,
      });
      position = 0;
    }

    equity = cash + position * price;
    equityCurve.push(equity / initialCapital);
  }

  const finalEquity = cash + position * (rows[rows.length - 1]!.close);
  const totalReturnPct = (finalEquity / initialCapital - 1) * 100;

  let peak = 1;
  let maxDrawdownPct = 0;
  for (const e of equityCurve) {
    if (e > peak) peak = e;
    const dd = (1 - e / peak) * 100;
    if (dd > maxDrawdownPct) maxDrawdownPct = dd;
  }

  const days = rows.length;
  const years = Math.max(0.001, days / 365);
  const annualizedReturnPct = (Math.pow(finalEquity / initialCapital, 1 / years) - 1) * 100;

  const sellTrades = trades.filter((t) => t.side === "sell" && t.pnl != null);
  const wins = sellTrades.filter((t) => (t.pnl ?? 0) > 0).length;
  const winRate = sellTrades.length > 0 ? wins / sellTrades.length : 0;

  return {
    total_return_pct: Math.round(totalReturnPct * 100) / 100,
    annualized_return_pct: Math.round(annualizedReturnPct * 100) / 100,
    max_drawdown_pct: Math.round(maxDrawdownPct * 100) / 100,
    trade_count: trades.length,
    win_rate: Math.round(winRate * 100) / 100,
    equity_curve: equityCurve.map((x) => Math.round(x * 1e6) / 1e6),
    trades,
    start_date,
    end_date,
  };
}

File: apps/backtest-python/engine.py
"""
백테스트 엔진 — pandas, yfinance 기반.
히스토리 조회 + RSI 임계값 전략 시뮬레이션.
"""
from __future__ import annotations

import pandas as pd

try:
    import yfinance as yf
except ImportError:
    yf = None


def get_history(symbol: str, start: str, end: str, interval: str = "1d") -> list[dict]:
    """yfinance로 일봉 조회. 실패 시 빈 리스트."""
    if not yf:
        return []
    try:
        ticker = yf.Ticker(symbol)
        df = ticker.history(start=start, end=end, interval=interval or "1d")
        if df.empty:
            return []
        df = df.reset_index()
        df["Date"] = df["Date"].dt.strftime("%Y-%m-%d")
        return df[["Date", "Open", "High", "Low", "Close", "Volume"]].rename(
            columns={"Date": "date", "Open": "open", "High": "high", "Low": "low", "Close": "close", "Volume": "volume"}
        ).to_dict(orient="records")
    except Exception:
        return []


def compute_rsi(series: pd.Series, period: int = 14) -> pd.Series:
    """close 시리즈로 RSI(period) 계산."""
    delta = series.diff()
    gain = delta.clip(lower=0)
    loss = (-delta).clip(lower=0)
    avg_gain = gain.rolling(period).mean()
    avg_loss = loss.rolling(period).mean()
    rs = avg_gain / avg_loss.replace(0, float("nan"))
    rsi = 100 - (100 / (1 + rs))
    return rsi


def run_backtest(
    start_date: str,
    end_date: str,
    symbols: list[str],
    strategy_type: str = "rsi_threshold",
    period: int = 14,
    buy_below: float = 30.0,
    sell_above: float = 70.0,
    initial_capital: float = 1_000_000.0,
) -> dict:
    """RSI 임계값 전략 백테스트. Node 엔진과 동일한 응답 형식."""
    symbol = symbols[0] if symbols else "AAPL"
    rows = get_history(symbol, start_date, end_date)
    if not rows:
        return {
            "total_return_pct": 0.0,
            "annualized_return_pct": 0.0,
            "max_drawdown_pct": 0.0,
            "trade_count": 0,
            "win_rate": 0.0,
            "equity_curve": [],
            "trades": [],
            "start_date": start_date,
            "end_date": end_date,
        }

    df = pd.DataFrame(rows)
    df["close"] = pd.to_numeric(df["close"], errors="coerce").ffill()
    df = df.dropna(subset=["close"])
    if df.empty or len(df) < period + 1:
        return {
            "total_return_pct": 0.0,
            "annualized_return_pct": 0.0,
            "max_drawdown_pct": 0.0,
            "trade_count": 0,
            "win_rate": 0.0,
            "equity_curve": [1.0] * len(df),
            "trades": [],
            "start_date": start_date,
            "end_date": end_date,
        }

    df["rsi"] = compute_rsi(df["close"], period)
    df = df.dropna(subset=["rsi"])

    trades: list[dict] = []
    position = 0
    entry_price = 0.0
    cash = initial_capital
    equity_curve: list[float] = []

    for _, row in df.iterrows():
        date = row["date"]
        price = float(row["close"])
        rsi = float(row["rsi"])

        if position == 0 and rsi < buy_below:
            qty = int(cash / price)
            if qty > 0:
                position = qty
                entry_price = price
                cash -= price * qty
                trades.append({"date": date, "symbol": symbol, "side": "buy", "price": price, "quantity": qty})
        elif position > 0 and rsi > sell_above:
            pnl = (price - entry_price) * position
            pnl_pct = (price / entry_price - 1) * 100
            cash += price * position
            trades.append({
                "date": date,
                "symbol": symbol,
                "side": "sell",
                "price": price,
                "quantity": position,
                "pnl": round(pnl, 2),
                "pnl_pct": round(pnl_pct, 2),
            })
            position = 0

        equity = cash + position * price
        equity_curve.append(round(equity / initial_capital, 6))

    final_equity = cash + position * float(df.iloc[-1]["close"])
    total_return_pct = (final_equity / initial_capital - 1) * 100
    years = max(0.001, len(df) / 365.0)
    annualized_return_pct = (pow(final_equity / initial_capital, 1 / years) - 1) * 100

    peak = 1.0
    max_drawdown_pct = 0.0
    for e in equity_curve:
        if e > peak:
            peak = e
        dd = (1 - e / peak) * 100
        if dd > max_drawdown_pct:
            max_drawdown_pct = dd

    sell_trades = [t for t in trades if t["side"] == "sell" and "pnl" in t]
    wins = sum(1 for t in sell_trades if t["pnl"] > 0)
    win_rate = (wins / len(sell_trades)) if sell_trades else 0.0

    return {
        "total_return_pct": round(total_return_pct, 2),
        "annualized_return_pct": round(annualized_return_pct, 2),
        "max_drawdown_pct": round(max_drawdown_pct, 2),
        "trade_count": len(trades),
        "win_rate": round(win_rate, 2),
        "equity_curve": equity_curve,
        "trades": trades,
        "start_date": start_date,
        "end_date": end_date,
    }
